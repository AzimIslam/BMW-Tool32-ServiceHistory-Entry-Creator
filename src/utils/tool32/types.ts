import { ServiceItem, MaintenanceStatus } from './enums'

export type ServiceEntry = {
  item: ServiceItem
  status: MaintenanceStatus
  distanceRemaining: number
  monthsRemaining: number
}
