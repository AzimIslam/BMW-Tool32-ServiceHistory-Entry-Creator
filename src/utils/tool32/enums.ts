export enum MaintenanceStatus {
  ON_TIME = '0x1',
  LATE = '0x2',
  INCOMPLETE = '0x3'
}

export enum MileageUnit {
  MI,
  KM
}

export enum BMWService {
  NO = 0,
  YES = 1
}

export enum ServiceItem {
  ENGINE_OIL = 1,
  FRONT_BRAKES = 2,
  BRAKE_FLUID = 3,
  MICROFILTER = 4,
  RECIRCULATION_FILTER = 5,
  REAR_BRAKES = 6,
  SPARK_PLUGS = 10,
  AIR_FILTER_ELEMENT = 11,
  FUEL_FILTER = 12,
  STATUTORY_VEHICLE_INSPECTION = 32,
  VEHICLE_INSPECTION = 100
}
