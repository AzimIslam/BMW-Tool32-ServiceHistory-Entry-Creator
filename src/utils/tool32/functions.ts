import { BMWService, MileageUnit, MaintenanceStatus } from './enums'
import type { ServiceEntry } from './types'

import InvalidArgumentException from './exceptions/InvalidArgumentException'

export function createServiceHistoryEntryArg(
  date: Date,
  mileage: number,
  mileageUnit: MileageUnit,
  serviceType: BMWService,
  dealerName: string,
  maintenanceStatus: MaintenanceStatus,
  services: ServiceEntry[]
): string {
  if (services.length == 0)
    throw new InvalidArgumentException('Cannot generate argument, if there are no service entries')

  const day: string = padZero(date.getDate()).toString()
  const month: string = padZero(date.getMonth() + 1).toString() // Months are zero-based
  const year: string = date.getFullYear().toString()
  const actualMileage = mileageUnit == MileageUnit.MI ? miToKm(mileage) : mileage

  let argument = `${day};${month};${year};0;${actualMileage};${serviceType.toString()};${dealerName};${maintenanceStatus.toString()};${services.length}`

  for (const { item, status, distanceRemaining, monthsRemaining } of services) {
    if (status == MaintenanceStatus.ON_TIME && distanceRemaining != 0 && monthsRemaining != 0)
      throw new InvalidArgumentException(
        'Illegal state, cannot service on time and have remaining distance or mileage left.'
      )
    const actualDistance =
      mileageUnit == MileageUnit.MI ? miToKm(distanceRemaining) : distanceRemaining
    argument += `;${item.toString()};${status.toString()};${actualDistance};${monthsRemaining};`
  }

  return argument
}

function padZero(num: number): string {
  return num < 10 ? `0${num}` : `${num}`
}

function miToKm(mi: number): number {
  if (mi == 0) return 0
  if (mi < 0) throw new InvalidArgumentException('Miles cannot be zero or negative')
  return Math.round(mi / 0.621371)
}
