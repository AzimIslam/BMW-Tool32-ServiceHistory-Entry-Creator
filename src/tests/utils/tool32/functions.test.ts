import { expect, it, describe, beforeAll } from 'vitest'
import { BMWService, MaintenanceStatus, MileageUnit, ServiceItem } from '@/utils/tool32/enums'
import type { ServiceEntry } from '../../../utils/tool32/types'
import { createServiceHistoryEntryArg } from '@/utils/tool32/functions'
import InvalidArgumentException from '@/utils/tool32/exceptions/InvalidArgumentException'

describe('#createServiceHistoryEntryArg', () => {
  const testDate = new Date(2024, 3, 13)
  const testMileage = 40000
  const testServiceType = BMWService.NO
  const testDealerName = 'BMAUTOT'
  const testMaintenanceStatus = MaintenanceStatus.ON_TIME

  it('should generate an argument successfully', () => {
    const testMileageUnit = MileageUnit.KM
    const testServices: ServiceEntry[] = [
      {
        item: ServiceItem.ENGINE_OIL,
        status: MaintenanceStatus.ON_TIME,
        distanceRemaining: 0,
        monthsRemaining: 0
      }
    ]

    const expectedResult = '13;04;2024;0;40000;0;BMAUTOT;0x1;1;1;0x1;0;0;'

    const actualResult = createServiceHistoryEntryArg(
      testDate,
      testMileage,
      testMileageUnit,
      testServiceType,
      testDealerName,
      testMaintenanceStatus,
      testServices
    )

    expect(actualResult).toEqual(expectedResult)
  })

  it('should convert miles to kilometres', () => {
    const testMileageUnit = MileageUnit.MI
    const testServices: ServiceEntry[] = [
      {
        item: ServiceItem.ENGINE_OIL,
        status: MaintenanceStatus.ON_TIME,
        distanceRemaining: 0,
        monthsRemaining: 0
      }
    ]

    const expectedResult = '13;04;2024;0;64374;0;BMAUTOT;0x1;1;1;0x1;0;0;'

    const actualResult = createServiceHistoryEntryArg(
      testDate,
      testMileage,
      testMileageUnit,
      testServiceType,
      testDealerName,
      testMaintenanceStatus,
      testServices
    )

    expect(actualResult).toEqual(expectedResult)
  })

  it('should throw InvalidArgumentException when no services are provided', () => {
    const testServices: ServiceEntry[] = []

    expect(() =>
      createServiceHistoryEntryArg(
        testDate,
        testMileage,
        MileageUnit.KM,
        testServiceType,
        testDealerName,
        testMaintenanceStatus,
        testServices
      )
    ).toThrowErrorMatchingSnapshot(
      'InvalidArgumentException: Cannot generate argument, if there are no service entries'
    )
  })

  it('should throw InvalidArgumentException when an entry has on time service with remaining distance or mileage', () => {
    const testServices: ServiceEntry[] = [
      {
        item: ServiceItem.ENGINE_OIL,
        status: MaintenanceStatus.ON_TIME,
        distanceRemaining: 100,
        monthsRemaining: 1
      }
    ]

    expect(() =>
      createServiceHistoryEntryArg(
        testDate,
        testMileage,
        MileageUnit.KM,
        testServiceType,
        testDealerName,
        testMaintenanceStatus,
        testServices
      )
    ).toThrowErrorMatchingSnapshot(
      'InvalidArgumentException: Cannot generate argument, if there are no service entries'
    )
  })
})
