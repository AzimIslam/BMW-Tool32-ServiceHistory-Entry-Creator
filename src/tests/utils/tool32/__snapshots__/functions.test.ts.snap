// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#createServiceHistoryEntryArg > should throw InvalidArgumentException when an entry has on time service with remaining distance or mileage > InvalidArgumentException: Cannot generate argument, if there are no service entries 1`] = `[InvalidArgumentException: Illegal state, cannot service on time and have remaining distance or mileage left.]`;

exports[`#createServiceHistoryEntryArg > should throw InvalidArgumentException when no services are provided > InvalidArgumentException: Cannot generate argument, if there are no service entries 1`] = `[InvalidArgumentException: Cannot generate argument, if there are no service entries]`;
